<h1>Содержание:<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Тема:-Анализ-базы-данных-электронной-библиотеки." data-toc-modified-id="Тема:-Анализ-базы-данных-электронной-библиотеки.-0">Тема: Анализ базы данных электронной библиотеки.</a></span></li><li><span><a href="#Цель:" data-toc-modified-id="Цель:-1">Цель:</a></span></li><li><span><a href="#Задачи:" data-toc-modified-id="Задачи:-2">Задачи:</a></span></li><li><span><a href="#Объект-исследования:" data-toc-modified-id="Объект-исследования:-3">Объект исследования:</a></span></li><li><span><a href="#1.-Изучение-данных-из-файлов" data-toc-modified-id="1.-Изучение-данных-из-файлов-4">1. Изучение данных из файлов</a></span><ul class="toc-item"><li><span><a href="#1.1.-Импортирование-библиотек" data-toc-modified-id="1.1.-Импортирование-библиотек-4.1">1.1. Импортирование библиотек</a></span></li><li><span><a href="#1.2.-Загрузка-данных" data-toc-modified-id="1.2.-Загрузка-данных-4.2">1.2. Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#1.2.1.-Подключение-базы-данных" data-toc-modified-id="1.2.1.-Подключение-базы-данных-4.2.1">1.2.1. Подключение базы данных</a></span></li><li><span><a href="#1.2.2.-Чтение-файлов" data-toc-modified-id="1.2.2.-Чтение-файлов-4.2.2">1.2.2. Чтение файлов</a></span></li></ul></li><li><span><a href="#1.3.-Предварительное-изучение-данных" data-toc-modified-id="1.3.-Предварительное-изучение-данных-4.3">1.3. Предварительное изучение данных</a></span><ul class="toc-item"><li><span><a href="#1.3.1.-Данные-о-книгах" data-toc-modified-id="1.3.1.-Данные-о-книгах-4.3.1">1.3.1. Данные о книгах</a></span></li><li><span><a href="#1.3.2.-Данные-об-авторах" data-toc-modified-id="1.3.2.-Данные-об-авторах-4.3.2">1.3.2. Данные об авторах</a></span></li><li><span><a href="#1.3.3.-Данные-об-издательствах" data-toc-modified-id="1.3.3.-Данные-об-издательствах-4.3.3">1.3.3. Данные об издательствах</a></span></li><li><span><a href="#1.3.4.-Данные-о-пользовательских-оценках" data-toc-modified-id="1.3.4.-Данные-о-пользовательских-оценках-4.3.4">1.3.4. Данные о пользовательских оценках</a></span></li><li><span><a href="#1.3.5.-Данные-о-пользовательских-обзорах-на-книги" data-toc-modified-id="1.3.5.-Данные-о-пользовательских-обзорах-на-книги-4.3.5">1.3.5. Данные о пользовательских обзорах на книги</a></span></li></ul></li></ul></li><li><span><a href="#2.-SQL--запросы" data-toc-modified-id="2.-SQL--запросы-5">2. SQL- запросы</a></span><ul class="toc-item"><li><span><a href="#2.1.-Количество-книг,-вышедших-после-01.01.2020" data-toc-modified-id="2.1.-Количество-книг,-вышедших-после-01.01.2020-5.1">2.1. Количество книг, вышедших после 01.01.2020</a></span></li><li><span><a href="#2.2.-Количество-обзоров-и-средняя-оценка-для-каждой-книги" data-toc-modified-id="2.2.-Количество-обзоров-и-средняя-оценка-для-каждой-книги-5.2">2.2. Количество обзоров и средняя оценка для каждой книги</a></span></li><li><span><a href="#2.3.-Издательство,-выпустившее-наибольшее-количество-книг-толщиной-более-50-страниц" data-toc-modified-id="2.3.-Издательство,-выпустившее-наибольшее-количество-книг-толщиной-более-50-страниц-5.3">2.3. Издательство, выпустившее наибольшее количество книг толщиной более 50 страниц</a></span></li><li><span><a href="#2.4.-Автор-с-самой-высокой-средней-оценкой-книг" data-toc-modified-id="2.4.-Автор-с-самой-высокой-средней-оценкой-книг-5.4">2.4. Автор с самой высокой средней оценкой книг</a></span></li><li><span><a href="#2.5.-Среднее-количество-обзоров-от-пользователей,-которые-поставили-более-50-оценок" data-toc-modified-id="2.5.-Среднее-количество-обзоров-от-пользователей,-которые-поставили-более-50-оценок-5.5">2.5. Среднее количество обзоров от пользователей, которые поставили более 50 оценок</a></span></li></ul></li><li><span><a href="#3.-Общий-вывод" data-toc-modified-id="3.-Общий-вывод-6">3. Общий вывод</a></span></li></ul></div>

## Тема: Анализ базы данных электронной библиотеки.

## Цель:
**Исследовать базу данных сервиса для чтения книг по подписке при помощи SQL- запросов.**

## Задачи:
**1. Изучить файлы с данными.**

**2. Посчитать количество книг, вышедших после 01.01.2020.**

**3. Посчитать количество обзоров и среднюю оценку для каждой книги.**

**4. Определить издательство, выпустившее наибольшее количество книг толщиной более 50 страниц (в целях исключения брошюр).**

**5. Установить автора с самой высокой средней оценкой книг (учитывать книги с 50 и более оценками).**

**6. Посчитать среднее количество обзоров от пользователей, которые поставили более 50 оценок.**

**7. Сформулировать основные выводы.**


## Объект исследования: 
**Таблицы с данными:**
- О книгах.
- Об авторах.
- Об издательствах.
- О пользовательских оценках книг.
- О пользовательских обзорах на книги.

***Заказчик:***  Владельцы крупного сервиса для чтения книг по подписке (электронная библиотека).

## 1. Изучение данных из файлов

### 1.1. Импортирование библиотек

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
from pandas.io import sql
import sqlite3

### 1.2. Загрузка данных

#### 1.2.1. Подключение базы данных

In [2]:
# Устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# Сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

#### 1.2.2. Чтение файлов

In [3]:
# Данные о книгах
query_1 = ''' SELECT *
            FROM books
        '''
books = pd.io.sql.read_sql(query_1, con = engine) 

# Данные об авторах
query_2 = ''' SELECT *
            FROM authors
        '''
authors = pd.io.sql.read_sql(query_2, con = engine) 

# Данные об издательствах
query_3 = ''' SELECT *
            FROM publishers
        '''
publishers = pd.io.sql.read_sql(query_3, con = engine) 

# Данные об оценках пользователей
query_4 = ''' SELECT *
            FROM ratings
        '''
ratings = pd.io.sql.read_sql(query_4, con = engine) 

# Данные об обзорах на книги
query_5 = ''' SELECT *
            FROM reviews
        '''
reviews = pd.io.sql.read_sql(query_5, con = engine) 

### 1.3. Предварительное изучение данных

In [4]:
# Напишем функцию, возвращающую по sql запросу 2 строки таблицы
def sql(query):
    return pd.io.sql.read_sql(query, con = engine)
def  df_view(table):
    result = sql(f'''
        SELECT
            *
        FROM
            {table}
        LIMIT
            2
        ''')
    return result

In [5]:
# Напишем функцию, возвращающую количество дубликатов и общую информацию о таблице
def df_viewing(df):
    display(df.duplicated().sum()) 
    display(df.info()) 

#### 1.3.1. Данные о книгах

In [6]:
df_view('books')

book_id  author_id                               title  num_pages  \
0        1        546                        'Salem's Lot        594   
1        2        465  1 000 Places to See Before You Die        992   

  publication_date  publisher_id  
0       2005-11-01            93  
1       2003-05-22           336

In [7]:
df_viewing(books)

0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
book_id             1000 non-null int64
author_id           1000 non-null int64
title               1000 non-null object
num_pages           1000 non-null int64
publication_date    1000 non-null object
publisher_id        1000 non-null int64
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

**В таблице с данными о книгах, 6 столбцов, 1 000 строк. Пропуски данных и дубликаты отсутствуют.**

**Столбцы содержат следующую информацию:**
- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

#### 1.3.2. Данные об авторах

In [8]:
df_view('authors')

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs

In [9]:
df_viewing(authors)

0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
author_id    636 non-null int64
author       636 non-null object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

**В таблице с данными об авторах, 2 столбца, 636 строк. Пропуски данных и дубликаты отсутствуют.**

**Столбцы содержат следующую информацию:**
- `author_id` — идентификатор автора;
- `author` — имя автора.

#### 1.3.3. Данные об издательствах

In [10]:
df_view('publishers')

publisher_id publisher
0             1       Ace
1             2  Ace Book

In [11]:
df_viewing(publishers)

0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
publisher_id    340 non-null int64
publisher       340 non-null object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

**В таблице с данными об издательствах, 2 столбца, 340 строк. Пропуски данных и дубликаты отсутствуют.**

**Столбцы содержат следующую информацию:**
- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства.

#### 1.3.4. Данные о пользовательских оценках

In [12]:
df_view('ratings')

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2

In [13]:
df_viewing(ratings)

0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
rating_id    6456 non-null int64
book_id      6456 non-null int64
username     6456 non-null object
rating       6456 non-null int64
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

**В таблице с данными о пользовательских оценках, 4 столбца, 6456 строк. Пропуски данных и дубликаты отсутствуют.**

**Столбцы содержат следующую информацию:**
- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

#### 1.3.5. Данные о пользовательских обзорах на книги

In [14]:
df_view('reviews')

review_id  book_id      username  \
0          1        1  brandtandrea   
1          2        1    ryanfranco   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...

In [15]:
df_viewing(reviews)

0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
review_id    2793 non-null int64
book_id      2793 non-null int64
username     2793 non-null object
text         2793 non-null object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

**В таблице с данными о пользовательских обзорах на книги, 4 столбца, 2793 строки. Пропуски данных и дубликаты отсутствуют.**

**Столбцы содержат следующую информацию:**
- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

**Полученные из базы данные корректны. Можно приступать к SQL- запросам.**

## 2. SQL- запросы

### 2.1. Количество книг, вышедших после 01.01.2020

In [16]:
query_6 = ''' SELECT
                    COUNT(book_id) AS cnt_book_2000
                FROM
                    books
                WHERE
                   publication_date > '2000-01-01' 
            '''
book_2000 = pd.io.sql.read_sql(query_6, con = engine) 

book_2000

cnt_book_2000
0            819

**После 01.01.2000 выпущено 819 книг.**

### 2.2. Количество обзоров и средняя оценка для каждой книги

In [17]:
query_7 = ''' SELECT
                authors.author AS author, 
                books.title, 
                COUNT(DISTINCT reviews.review_id) AS cnt_reviews_book, 
                AVG (ratings.rating) AS median_rating  
            FROM
                authors
                INNER JOIN books ON books.author_id  = authors.author_id
                LEFT JOIN reviews ON books.book_id  = reviews.book_id
                INNER JOIN ratings ON books.book_id  = ratings.book_id
            
            GROUP BY
                  authors.author, 
                books.title 
            ORDER BY 
                authors.author 
            '''
overview = pd.io.sql.read_sql(query_7, con = engine) 

overview

author  \
0                              A.S. Byatt   
1          Aesop/Laura Harris/Laura Gibbs   
2                         Agatha Christie   
3                         Agatha Christie   
4                         Agatha Christie   
..                                    ...   
994                           Zadie Smith   
995                           Zadie Smith   
996                 Zilpha Keatley Snyder   
997                    Zora Neale Hurston   
998  Åsne Seierstad/Ingrid Christopherson   

                                        title  cnt_reviews_book  median_rating  
0                                  Possession                 3       4.000000  
1                              Aesop's Fables                 2       3.750000  
2     A Murder Is Announced (Miss Marple  #5)                 2       4.000000  
3                    And Then There Were None                 5       4.454545  
4    Evil Under the Sun (Hercule Poirot  #24)                 2       5.000000  
..                                        ...               ...            ...  
994                                 On Beauty                 2       4.500000  
995                               White Teeth                 4       3.750000  
996                            The Egypt Game                 2       4.000000  
997              Their Eyes Were Watching God                 4       4.125000  
998                   The Bookseller of Kabul                 2       4.000000  

[999 rows x 4 columns]

Таблица содержит на 1 книгу меньше, чем по данным о книгах. Дополнительно посмотрим на число книг по ID и по названию.

In [18]:
query_7_1 = ''' SELECT
                COUNT(DISTINCT books.book_id) 
                
            FROM
                books
            '''
n1 = pd.io.sql.read_sql(query_7_1, con = engine) 

n1

count
0   1000

In [19]:
query_7_1 = ''' SELECT
                COUNT(DISTINCT books.title) 
                
            FROM
                books
            '''
n2 = pd.io.sql.read_sql(query_7_1, con = engine) 

n2

count
0    999

В таблице имеется одна книга с одинаковым названием, но разными ID.

**Получили 999 книг с отзывами (от 2 до 1120 на книгу) и средними оценками от 1,5 до 5.**

### 2.3. Издательство, выпустившее наибольшее количество книг толщиной более 50 страниц

In [20]:
query_8 = ''' SELECT
                publishers.publisher AS publishing_house, 
                COUNT(DISTINCT books.book_id) AS cnt_real_book 
            FROM
                publishers
                INNER JOIN books ON books.publisher_id  = publishers.publisher_id
            WHERE
                books.num_pages > 50
            GROUP BY
                  publishers.publisher 
            ORDER BY  
                cnt_real_book DESC
            '''
publishing_house = pd.io.sql.read_sql(query_8, con = engine) 

publishing_house

publishing_house  cnt_real_book
0                          Penguin Books             42
1                                Vintage             31
2               Grand Central Publishing             25
3                       Penguin Classics             24
4                                 Bantam             19
..                                   ...            ...
329                  Random House Anchor              1
330  Random House Audio Publishing Group              1
331  Random House Large Print Publishing              1
332                   Random House Trade              1
333                 Random House Vintage              1

[334 rows x 2 columns]

**"Penguin Books" - издательство, выпустившее наибольшее количество книг (42), толщиной более 50 страниц.**

### 2.4. Автор с самой высокой средней оценкой книг

In [21]:
query_9 = ''' SELECT
                authors.author AS author, 
                COUNT(DISTINCT ratings.rating_id) AS cnt_rating_book,
                AVG (ratings.rating) AS median_rating  
            FROM
                authors
                INNER JOIN books ON books.author_id  = authors.author_id
                INNER JOIN ratings ON books.book_id  = ratings.book_id
            GROUP BY
                  authors.author 
            HAVING
                COUNT(DISTINCT ratings.rating_id) > 50
            ORDER BY 
                median_rating DESC
            '''
author_rating = pd.io.sql.read_sql(query_9, con = engine) 

author_rating

author  cnt_rating_book  \
0                          J.K. Rowling/Mary GrandPré              312   
1                                     Agatha Christie               53   
2                   Markus Zusak/Cao Xuân Việt Khương               53   
3                                      J.R.R. Tolkien              166   
4                            Roald Dahl/Quentin Blake               62   
5                                   Louisa May Alcott               54   
6                                        Rick Riordan               84   
7                                       Arthur Golden               56   
8                                        Stephen King              106   
9                                        John Grisham               70   
10                                    William Golding               71   
11                                    Nicholas Sparks              111   
12                                       Jodi Picoult               76   
13                                    Sophie Kinsella               57   
14                                    James Patterson               64   
15                                      J.D. Salinger               98   
16           Paulo Coelho/Alan R. Clarke/Özdemir İnce               57   
17  William Shakespeare/Paul Werstine/Barbara A. M...               66   
18                                          Dan Brown              143   
19                                         Lois Lowry               65   
20            George Orwell/Boris Grabnar/Peter Škerl               74   
21                                    Stephenie Meyer              160   
22                                     John Steinbeck               73   

    median_rating  
0        4.288462  
1        4.283019  
2        4.264151  
3        4.240964  
4        4.209677  
5        4.203704  
6        4.130952  
7        4.107143  
8        4.009434  
9        3.971429  
10       3.901408  
11       3.882883  
12       3.881579  
13       3.877193  
14       3.859375  
15       3.846939  
16       3.789474  
17       3.787879  
18       3.741259  
19       3.738462  
20       3.729730  
21       3.662500  
22       3.643836

**Автор с самым высоким рейтингом, оцененный более, чем 50 пользователями - Джоан Роулинг, совместно с иллюстратором ее книг - Мэри Гранпрэ.**

### 2.5. Среднее количество обзоров от пользователей, которые поставили более 50 оценок

In [22]:
query_10 = ''' SELECT
                AVG(cnt_review)  as avg_review
            FROM 
            (SELECT
                reviews.username AS username, 
                COUNT(DISTINCT ratings.rating_id) AS cnt_rating_book,
                COUNT (DISTINCT reviews.review_id) AS cnt_review  
            FROM
                reviews
                INNER JOIN ratings ON ratings.username  = reviews.username
            GROUP BY
                  reviews.username 
            HAVING
                COUNT(DISTINCT ratings.rating_id) > 50
            )
            AS avg_review 
            '''
median_review  = pd.io.sql.read_sql(query_10, con = engine) 

median_review

avg_review
0   24.333333

## 3. Общий вывод

В рамках исследования проводилось изучение базы данных сервиса для чтения книг, а именно 5 таблиц с информацией о книгах, авторах, издательствах, пользовательких оценках и обзорах. 

**Основные запросы данных осуществлялись посредством SQL.**

В процессе проведения запросов установлено наличие одного дубля в данных о книгах - одинаковые названия, разные ID.

**В задании исследования были поставлены конкретные задачи, по итогам SQL- запросов получены следующие результаты.**

**1. В базе имеется 819 книг, выпущенных после 01.01.2000.** 

**2. В базе имеется 999 книг, которым сделаны обзоры и поставлены оценки. Число обзоров  варьируется от 2 до 1120 на одну книгу. Средняя оценка книги пользователями варьируется от 1,5 до 5. **

**3. Наибольшее количество книг (толщиной более 50 страниц, не брошюр) выпустило издательство "Penguin Books".**

**4. Наибольшее количество книг (толщиной более 50 страниц, не брошюр) выпустило издательство "Penguin Books". На втором месте издательство Vintage	 (31 книга), на третьем - Grand Central Publishing (25 книг), незначительно отстает издательство Penguin Classics (24 книги).**

**5. Из числа авторов, книги которых получили более 50 оценок, наиболее высокий рейтинг у следующих авторов:**
- Джоан Роулинг - рейтинг 4,289
- Агата Кристи - рейтинг 4,283
- Ма́ркус Зусак - рейтинг 4,264
- Джон Рональд Руэл Толкин - рейтинг 4,241
- Роальд Даль - рейтинг 4,21

**6. Пользователи, поставившие более 50 оценок, в среднем написали около 24 отзывов.**

***Рекомендации***
- Провести дополнительное исследование по жанрам книг, имеющихся в базе данных.
- Изучить аудиторию сервиса - регион, источник захода на сайт (устройство), предпочтения.